{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MUk9mUkhI8Je"
   },
   "source": [
    "Устанавливаем HMMER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "1E2ttARVSKuE",
    "outputId": "82f90b77-cd0f-4e1f-d934-80884cfdda3e"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--2025-06-18 03:34:49--  http://eddylab.org/software/hmmer/hmmer-3.3.2.tar.gz\n",
      "Resolving eddylab.org (eddylab.org)... 96.126.110.11, 2600:3c03::f03c:91ff:fec8:383c\n",
      "Connecting to eddylab.org (eddylab.org)|96.126.110.11|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 18213049 (17M) [application/x-gzip]\n",
      "Saving to: ‘hmmer-3.3.2.tar.gz’\n",
      "\n",
      "     0K .......... .......... .......... .......... ..........  0% 84,5K 3m30s\n",
      "    50K .......... .......... .......... .......... ..........  0%  249K 2m20s\n",
      "   100K .......... .......... .......... .......... ..........  0% 5,95M 94s\n",
      "   150K .......... .......... .......... .......... ..........  1%  232K 89s\n",
      "   200K .......... .......... .......... .......... ..........  1%  299K 83s\n",
      "   250K .......... .......... .......... .......... ..........  1% 1,83M 71s\n",
      "   300K .......... .......... .......... .......... ..........  1%  248K 70s\n",
      "   350K .......... .......... .......... .......... ..........  2% 3,23M 62s\n",
      "   400K .......... .......... .......... .......... ..........  2%  243K 63s\n",
      "   450K .......... .......... .......... .......... ..........  2%  187K 66s\n",
      "   500K .......... .......... .......... .......... ..........  3% 4,71M 60s\n",
      "   550K .......... .......... .......... .......... ..........  3%  232K 61s\n",
      "   600K .......... .......... .......... .......... ..........  3% 6,43M 56s\n",
      "   650K .......... .......... .......... .......... ..........  3%  183K 59s\n",
      "   700K .......... .......... .......... .......... ..........  4% 8,18M 55s\n",
      "   750K .......... .......... .......... .......... ..........  4%  249K 56s\n",
      "   800K .......... .......... .......... .......... ..........  4% 16,3M 52s\n",
      "   850K .......... .......... .......... .......... ..........  5%  251K 53s\n",
      "   900K .......... .......... .......... .......... ..........  5% 13,0M 50s\n",
      "   950K .......... .......... .......... .......... ..........  5%  215K 51s\n",
      "  1000K .......... .......... .......... .......... ..........  5% 8,29M 49s\n",
      "  1050K .......... .......... .......... .......... ..........  6%  188K 50s\n",
      "  1100K .......... .......... .......... .......... ..........  6% 10,6M 48s\n",
      "  1150K .......... .......... .......... .......... ..........  6% 8,83M 46s\n",
      "  1200K .......... .......... .......... .......... ..........  7%  211K 47s\n",
      "  1250K .......... .......... .......... .......... ..........  7% 6,97M 45s\n",
      "  1300K .......... .......... .......... .......... ..........  7%  245K 46s\n",
      "  1350K .......... .......... .......... .......... ..........  7% 12,3M 44s\n",
      "  1400K .......... .......... .......... .......... ..........  8% 10,8M 43s\n",
      "  1450K .......... .......... .......... .......... ..........  8%  196K 44s\n",
      "  1500K .......... .......... .......... .......... ..........  8% 4,10M 43s\n",
      "  1550K .......... .......... .......... .......... ..........  8% 10,4M 41s\n",
      "  1600K .......... .......... .......... .......... ..........  9%  209K 42s\n",
      "  1650K .......... .......... .......... .......... ..........  9% 16,1M 41s\n",
      "  1700K .......... .......... .......... .......... ..........  9% 2,14M 40s\n",
      "  1750K .......... .......... .......... .......... .......... 10%  231K 40s\n",
      "  1800K .......... .......... .......... .......... .......... 10% 3,47M 39s\n",
      "  1850K .......... .......... .......... .......... .......... 10% 7,63M 38s\n",
      "  1900K .......... .......... .......... .......... .......... 10%  275K 39s\n",
      "  1950K .......... .......... .......... .......... .......... 11% 1,83M 38s\n",
      "  2000K .......... .......... .......... .......... .......... 11% 5,22M 37s\n",
      "  2050K .......... .......... .......... .......... .......... 11%  326K 37s\n",
      "  2100K .......... .......... .......... .......... .......... 12% 1003K 36s\n",
      "  2150K .......... .......... .......... .......... .......... 12%  963K 36s\n",
      "  2200K .......... .......... .......... .......... .......... 12% 2,06M 35s\n",
      "  2250K .......... .......... .......... .......... .......... 12%  520K 35s\n",
      "  2300K .......... .......... .......... .......... .......... 13% 2,33M 34s\n",
      "  2350K .......... .......... .......... .......... .......... 13% 1,42M 33s\n",
      "  2400K .......... .......... .......... .......... .......... 13%  905K 33s\n",
      "  2450K .......... .......... .......... .......... .......... 14%  757K 33s\n",
      "  2500K .......... .......... .......... .......... .......... 14% 1,31M 32s\n",
      "  2550K .......... .......... .......... .......... .......... 14% 2,19M 32s\n",
      "  2600K .......... .......... .......... .......... .......... 14% 1,30M 31s\n",
      "  2650K .......... .......... .......... .......... .......... 15%  712K 31s\n",
      "  2700K .......... .......... .......... .......... .......... 15% 1,20M 30s\n",
      "  2750K .......... .......... .......... .......... .......... 15% 1,50M 30s\n",
      "  2800K .......... .......... .......... .......... .......... 16% 3,11M 29s\n",
      "  2850K .......... .......... .......... .......... .......... 16%  652K 29s\n",
      "  2900K .......... .......... .......... .......... .......... 16% 1,50M 29s\n",
      "  2950K .......... .......... .......... .......... .......... 16% 1,24M 28s\n",
      "  3000K .......... .......... .......... .......... .......... 17% 3,45M 28s\n",
      "  3050K .......... .......... .......... .......... .......... 17%  628K 28s\n",
      "  3100K .......... .......... .......... .......... .......... 17% 1,90M 27s\n",
      "  3150K .......... .......... .......... .......... .......... 17% 1,70M 27s\n",
      "  3200K .......... .......... .......... .......... .......... 18% 1,55M 26s\n",
      "  3250K .......... .......... .......... .......... .......... 18%  985K 26s\n",
      "  3300K .......... .......... .......... .......... .......... 18%  977K 26s\n",
      "  3350K .......... .......... .......... .......... .......... 19% 2,44M 26s\n",
      "  3400K .......... .......... .......... .......... .......... 19%  981K 25s\n",
      "  3450K .......... .......... .......... .......... .......... 19% 7,72M 25s\n",
      "  3500K .......... .......... .......... .......... .......... 19%  997K 25s\n",
      "  3550K .......... .......... .......... .......... .......... 20%  692K 25s\n",
      "  3600K .......... .......... .......... .......... .......... 20% 6,83M 24s\n",
      "  3650K .......... .......... .......... .......... .......... 20% 1,71M 24s\n",
      "  3700K .......... .......... .......... .......... .......... 21% 2,92M 23s\n",
      "  3750K .......... .......... .......... .......... .......... 21%  980K 23s\n",
      "  3800K .......... .......... .......... .......... .......... 21%  667K 23s\n",
      "  3850K .......... .......... .......... .......... .......... 21% 2,35M 23s\n",
      "  3900K .......... .......... .......... .......... .......... 22% 9,06M 23s\n",
      "  3950K .......... .......... .......... .......... .......... 22% 2,14M 22s\n",
      "  4000K .......... .......... .......... .......... .......... 22% 1,38M 22s\n",
      "  4050K .......... .......... .......... .......... .......... 23%  604K 22s\n",
      "  4100K .......... .......... .......... .......... .......... 23% 2,13M 22s\n",
      "  4150K .......... .......... .......... .......... .......... 23% 2,38M 21s\n",
      "  4200K .......... .......... .......... .......... .......... 23% 4,75M 21s\n",
      "  4250K .......... .......... .......... .......... .......... 24% 1,61M 21s\n",
      "  4300K .......... .......... .......... .......... .......... 24%  882K 21s\n",
      "  4350K .......... .......... .......... .......... .......... 24% 1,30M 21s\n",
      "  4400K .......... .......... .......... .......... .......... 25%  797K 20s\n",
      "  4450K .......... .......... .......... .......... .......... 25% 4,88M 20s\n",
      "  4500K .......... .......... .......... .......... .......... 25% 3,70M 20s\n",
      "  4550K .......... .......... .......... .......... .......... 25% 4,87M 20s\n",
      "  4600K .......... .......... .......... .......... .......... 26%  957K 19s\n",
      "  4650K .......... .......... .......... .......... .......... 26% 1,64M 19s\n",
      "  4700K .......... .......... .......... .......... .......... 26%  690K 19s\n",
      "  4750K .......... .......... .......... .......... .......... 26% 2,26M 19s\n",
      "  4800K .......... .......... .......... .......... .......... 27% 4,65M 19s\n",
      "  4850K .......... .......... .......... .......... .......... 27% 19,7M 18s\n",
      "  4900K .......... .......... .......... .......... .......... 27% 2,05M 18s\n",
      "  4950K .......... .......... .......... .......... .......... 28% 2,01M 18s\n",
      "  5000K .......... .......... .......... .......... .......... 28%  524K 18s\n",
      "  5050K .......... .......... .......... .......... .......... 28% 4,71M 18s\n",
      "  5100K .......... .......... .......... .......... .......... 28% 1,64M 18s\n",
      "  5150K .......... .......... .......... .......... .......... 29% 7,42M 17s\n",
      "  5200K .......... .......... .......... .......... .......... 29% 1,46M 17s\n",
      "  5250K .......... .......... .......... .......... .......... 29% 4,74M 17s\n",
      "  5300K .......... .......... .......... .......... .......... 30% 1,34M 17s\n",
      "  5350K .......... .......... .......... .......... .......... 30% 1,22M 17s\n",
      "  5400K .......... .......... .......... .......... .......... 30% 1,37M 17s\n",
      "  5450K .......... .......... .......... .......... .......... 30% 1,68M 17s\n",
      "  5500K .......... .......... .......... .......... .......... 31% 1,31M 16s\n",
      "  5550K .......... .......... .......... .......... .......... 31% 8,01M 16s\n",
      "  5600K .......... .......... .......... .......... .......... 31% 1,99M 16s\n",
      "  5650K .......... .......... .......... .......... .......... 32% 1,62M 16s\n",
      "  5700K .......... .......... .......... .......... .......... 32% 2,62M 16s\n",
      "  5750K .......... .......... .......... .......... .......... 32% 1,17M 16s\n",
      "  5800K .......... .......... .......... .......... .......... 32% 1,23M 16s\n",
      "  5850K .......... .......... .......... .......... .......... 33% 3,60M 15s\n",
      "  5900K .......... .......... .......... .......... .......... 33% 1,54M 15s\n",
      "  5950K .......... .......... .......... .......... .......... 33% 1,09M 15s\n",
      "  6000K .......... .......... .......... .......... .......... 34% 1,97M 15s\n",
      "  6050K .......... .......... .......... .......... .......... 34% 1,34M 15s\n",
      "  6100K .......... .......... .......... .......... .......... 34% 1,31M 15s\n",
      "  6150K .......... .......... .......... .......... .......... 34% 2,11M 15s\n",
      "  6200K .......... .......... .......... .......... .......... 35% 1,37M 14s\n",
      "  6250K .......... .......... .......... .......... .......... 35% 1,29M 14s\n",
      "  6300K .......... .......... .......... .......... .......... 35% 6,93M 14s\n",
      "  6350K .......... .......... .......... .......... .......... 35% 3,03M 14s\n",
      "  6400K .......... .......... .......... .......... .......... 36% 3,84M 14s\n",
      "  6450K .......... .......... .......... .......... .......... 36% 1,72M 14s\n",
      "  6500K .......... .......... .......... .......... .......... 36% 1,54M 14s\n",
      "  6550K .......... .......... .......... .......... .......... 37% 1,57M 14s\n",
      "  6600K .......... .......... .......... .......... .......... 37%  925K 14s\n",
      "  6650K .......... .......... .......... .......... .......... 37% 1000K 13s\n",
      "  6700K .......... .......... .......... .......... .......... 37% 2,36M 13s\n",
      "  6750K .......... .......... .......... .......... .......... 38% 2,07M 13s\n",
      "  6800K .......... .......... .......... .......... .......... 38% 1,60M 13s\n",
      "  6850K .......... .......... .......... .......... .......... 38% 1,88M 13s\n",
      "  6900K .......... .......... .......... .......... .......... 39% 2,70M 13s\n",
      "  6950K .......... .......... .......... .......... .......... 39% 1,74M 13s\n",
      "  7000K .......... .......... .......... .......... .......... 39%  750K 13s\n",
      "  7050K .......... .......... .......... .......... .......... 39% 5,42M 13s\n",
      "  7100K .......... .......... .......... .......... .......... 40% 1,88M 12s\n",
      "  7150K .......... .......... .......... .......... .......... 40% 3,00M 12s\n",
      "  7200K .......... .......... .......... .......... .......... 40%  941K 12s\n",
      "  7250K .......... .......... .......... .......... .......... 41% 3,55M 12s\n",
      "  7300K .......... .......... .......... .......... .......... 41% 1,24M 12s\n",
      "  7350K .......... .......... .......... .......... .......... 41% 1,63M 12s\n",
      "  7400K .......... .......... .......... .......... .......... 41%  879K 12s\n",
      "  7450K .......... .......... .......... .......... .......... 42% 6,78M 12s\n",
      "  7500K .......... .......... .......... .......... .......... 42%  111K 12s\n",
      "  7550K .......... .......... .......... .......... .......... 42%  241M 12s\n",
      "  7600K .......... .......... .......... .......... .......... 43%  214M 12s\n",
      "  7650K .......... .......... .......... .......... .......... 43%  559K 12s\n",
      "  7700K .......... .......... .......... .......... .......... 43%  169M 12s\n",
      "  7750K .......... .......... .......... .......... .......... 43%  246M 12s\n",
      "  7800K .......... .......... .......... .......... .......... 44%  274M 12s\n",
      "  7850K .......... .......... .......... .......... .......... 44%  260M 11s\n",
      "  7900K .......... .......... .......... .......... .......... 44%  225M 11s\n",
      "  7950K .......... .......... .......... .......... .......... 44%  269M 11s\n",
      "  8000K .......... .......... .......... .......... .......... 45%  244M 11s\n",
      "  8050K .......... .......... .......... .......... .......... 45%  151K 11s\n",
      "  8100K .......... .......... .......... .......... .......... 45% 1,32M 11s\n",
      "  8150K .......... .......... .......... .......... .......... 46%  227M 11s\n",
      "  8200K .......... .......... .......... .......... .......... 46%  265M 11s\n",
      "  8250K .......... .......... .......... .......... .......... 46%  596K 11s\n",
      "  8300K .......... .......... .......... .......... .......... 46%  179M 11s\n",
      "  8350K .......... .......... .......... .......... .......... 47%  261M 11s\n",
      "  8400K .......... .......... .......... .......... .......... 47%  265M 11s\n",
      "  8450K .......... .......... .......... .......... .......... 47% 1,34M 11s\n",
      "  8500K .......... .......... .......... .......... .......... 48% 1,42M 10s\n",
      "  8550K .......... .......... .......... .......... .......... 48% 1,21M 10s\n",
      "  8600K .......... .......... .......... .......... .......... 48%  981K 10s\n",
      "  8650K .......... .......... .......... .......... .......... 48%  988K 10s\n",
      "  8700K .......... .......... .......... .......... .......... 49% 1,37M 10s\n",
      "  8750K .......... .......... .......... .......... .......... 49%  886K 10s\n",
      "  8800K .......... .......... .......... .......... .......... 49% 4,15M 10s\n",
      "  8850K .......... .......... .......... .......... .......... 50% 1,21M 10s\n",
      "  8900K .......... .......... .......... .......... .......... 50% 1,29M 10s\n",
      "  8950K .......... .......... .......... .......... .......... 50%  570K 10s\n",
      "  9000K .......... .......... .......... .......... .......... 50% 2,00M 10s\n",
      "  9050K .......... .......... .......... .......... .......... 51%  828K 10s\n",
      "  9100K .......... .......... .......... .......... .......... 51% 2,67M 10s\n",
      "  9150K .......... .......... .......... .......... .......... 51% 2,64M 10s\n",
      "  9200K .......... .......... .......... .......... .......... 52%  626K 9s\n",
      "  9250K .......... .......... .......... .......... .......... 52% 1,23M 9s\n",
      "  9300K .......... .......... .......... .......... .......... 52%  704K 9s\n",
      "  9350K .......... .......... .......... .......... .......... 52% 8,66M 9s\n",
      "  9400K .......... .......... .......... .......... .......... 53%  931K 9s\n",
      "  9450K .......... .......... .......... .......... .......... 53% 1,01M 9s\n",
      "  9500K .......... .......... .......... .......... .......... 53% 2,27M 9s\n",
      "  9550K .......... .......... .......... .......... .......... 53%  882K 9s\n",
      "  9600K .......... .......... .......... .......... .......... 54% 2,89M 9s\n",
      "  9650K .......... .......... .......... .......... .......... 54%  583K 9s\n",
      "  9700K .......... .......... .......... .......... .......... 54% 3,29M 9s\n",
      "  9750K .......... .......... .......... .......... .......... 55% 3,54M 9s\n",
      "  9800K .......... .......... .......... .......... .......... 55% 1,40M 9s\n",
      "  9850K .......... .......... .......... .......... .......... 55% 1,77M 9s\n",
      "  9900K .......... .......... .......... .......... .......... 55%  533K 9s\n",
      "  9950K .......... .......... .......... .......... .......... 56% 2,85M 8s\n",
      " 10000K .......... .......... .......... .......... .......... 56% 2,07M 8s\n",
      " 10050K .......... .......... .......... .......... .......... 56% 12,5M 8s\n",
      " 10100K .......... .......... .......... .......... .......... 57% 1,71M 8s\n",
      " 10150K .......... .......... .......... .......... .......... 57%  454K 8s\n",
      " 10200K .......... .......... .......... .......... .......... 57% 1,60M 8s\n",
      " 10250K .......... .......... .......... .......... .......... 57% 4,90M 8s\n",
      " 10300K .......... .......... .......... .......... .......... 58% 4,51M 8s\n",
      " 10350K .......... .......... .......... .......... .......... 58% 3,01M 8s\n",
      " 10400K .......... .......... .......... .......... .......... 58%  441K 8s\n",
      " 10450K .......... .......... .......... .......... .......... 59% 1,20M 8s\n",
      " 10500K .......... .......... .......... .......... .......... 59% 1,71M 8s\n",
      " 10550K .......... .......... .......... .......... .......... 59% 5,83M 8s\n",
      " 10600K .......... .......... .......... .......... .......... 59% 4,96M 8s\n",
      " 10650K .......... .......... .......... .......... .......... 60%  602K 8s\n",
      " 10700K .......... .......... .......... .......... .......... 60%  660K 7s\n",
      " 10750K .......... .......... .......... .......... .......... 60% 1,93M 7s\n",
      " 10800K .......... .......... .......... .......... .......... 61% 3,98M 7s\n",
      " 10850K .......... .......... .......... .......... .......... 61% 2,17M 7s\n",
      " 10900K .......... .......... .......... .......... .......... 61% 2,16M 7s\n",
      " 10950K .......... .......... .......... .......... .......... 61%  596K 7s\n",
      " 11000K .......... .......... .......... .......... .......... 62%  766K 7s\n",
      " 11050K .......... .......... .......... .......... .......... 62% 1,36M 7s\n",
      " 11100K .......... .......... .......... .......... .......... 62% 2,61M 7s\n",
      " 11150K .......... .......... .......... .......... .......... 62% 1,20M 7s\n",
      " 11200K .......... .......... .......... .......... .......... 63% 3,96M 7s\n",
      " 11250K .......... .......... .......... .......... .......... 63%  795K 7s\n",
      " 11300K .......... .......... .......... .......... .......... 63%  711K 7s\n",
      " 11350K .......... .......... .......... .......... .......... 64% 2,27M 7s\n",
      " 11400K .......... .......... .......... .......... .......... 64% 2,22M 7s\n",
      " 11450K .......... .......... .......... .......... .......... 64% 2,60M 7s\n",
      " 11500K .......... .......... .......... .......... .......... 64% 1,23M 6s\n",
      " 11550K .......... .......... .......... .......... .......... 65%  813K 6s\n",
      " 11600K .......... .......... .......... .......... .......... 65% 1,71M 6s\n",
      " 11650K .......... .......... .......... .......... .......... 65% 1,34M 6s\n",
      " 11700K .......... .......... .......... .......... .......... 66% 1,96M 6s\n",
      " 11750K .......... .......... .......... .......... .......... 66% 2,94M 6s\n",
      " 11800K .......... .......... .......... .......... .......... 66% 1,48M 6s\n",
      " 11850K .......... .......... .......... .......... .......... 66%  854K 6s\n",
      " 11900K .......... .......... .......... .......... .......... 67%  925K 6s\n",
      " 11950K .......... .......... .......... .......... .......... 67% 3,04M 6s\n",
      " 12000K .......... .......... .......... .......... .......... 67% 2,58M 6s\n",
      " 12050K .......... .......... .......... .......... .......... 68% 2,10M 6s\n",
      " 12100K .......... .......... .......... .......... .......... 68% 1,25M 6s\n",
      " 12150K .......... .......... .......... .......... .......... 68%  621K 6s\n",
      " 12200K .......... .......... .......... .......... .......... 68% 2,01M 6s\n",
      " 12250K .......... .......... .......... .......... .......... 69% 2,15M 6s\n",
      " 12300K .......... .......... .......... .......... .......... 69% 1,99M 6s\n",
      " 12350K .......... .......... .......... .......... .......... 69% 3,02M 5s\n",
      " 12400K .......... .......... .......... .......... .......... 69% 2,74M 5s\n",
      " 12450K .......... .......... .......... .......... .......... 70%  526K 5s\n",
      " 12500K .......... .......... .......... .......... .......... 70% 1,58M 5s\n",
      " 12550K .......... .......... .......... .......... .......... 70% 1,92M 5s\n",
      " 12600K .......... .......... .......... .......... .......... 71% 2,15M 5s\n",
      " 12650K .......... .......... .......... .......... .......... 71% 3,67M 5s\n",
      " 12700K .......... .......... .......... .......... .......... 71%  697K 5s\n",
      " 12750K .......... .......... .......... .......... .......... 71% 1,03M 5s\n",
      " 12800K .......... .......... .......... .......... .......... 72% 1,81M 5s\n",
      " 12850K .......... .......... .......... .......... .......... 72% 1,71M 5s\n",
      " 12900K .......... .......... .......... .......... .......... 72% 2,92M 5s\n",
      " 12950K .......... .......... .......... .......... .......... 73%  942K 5s\n",
      " 13000K .......... .......... .......... .......... .......... 73%  763K 5s\n",
      " 13050K .......... .......... .......... .......... .......... 73% 3,42M 5s\n",
      " 13100K .......... .......... .......... .......... .......... 73% 2,86M 5s\n",
      " 13150K .......... .......... .......... .......... .......... 74% 1,77M 5s\n",
      " 13200K .......... .......... .......... .......... .......... 74% 2,16M 5s\n",
      " 13250K .......... .......... .......... .......... .......... 74% 1,44M 4s\n",
      " 13300K .......... .......... .......... .......... .......... 75%  595K 4s\n",
      " 13350K .......... .......... .......... .......... .......... 75% 2,85M 4s\n",
      " 13400K .......... .......... .......... .......... .......... 75% 2,78M 4s\n",
      " 13450K .......... .......... .......... .......... .......... 75% 2,33M 4s\n",
      " 13500K .......... .......... .......... .......... .......... 76% 2,91M 4s\n",
      " 13550K .......... .......... .......... .......... .......... 76%  495K 4s\n",
      " 13600K .......... .......... .......... .......... .......... 76% 7,85M 4s\n",
      " 13650K .......... .......... .......... .......... .......... 77% 1,66M 4s\n",
      " 13700K .......... .......... .......... .......... .......... 77% 3,44M 4s\n",
      " 13750K .......... .......... .......... .......... .......... 77% 2,16M 4s\n",
      " 13800K .......... .......... .......... .......... .......... 77%  864K 4s\n",
      " 13850K .......... .......... .......... .......... .......... 78%  989K 4s\n",
      " 13900K .......... .......... .......... .......... .......... 78% 1,97M 4s\n",
      " 13950K .......... .......... .......... .......... .......... 78% 2,20M 4s\n",
      " 14000K .......... .......... .......... .......... .......... 78% 3,63M 4s\n",
      " 14050K .......... .......... .......... .......... .......... 79%  976K 4s\n",
      " 14100K .......... .......... .......... .......... .......... 79%  971K 4s\n",
      " 14150K .......... .......... .......... .......... .......... 79% 1,97M 4s\n",
      " 14200K .......... .......... .......... .......... .......... 80% 1,85M 3s\n",
      " 14250K .......... .......... .......... .......... .......... 80% 2,31M 3s\n",
      " 14300K .......... .......... .......... .......... .......... 80% 3,35M 3s\n",
      " 14350K .......... .......... .......... .......... .......... 80%  684K 3s\n",
      " 14400K .......... .......... .......... .......... .......... 81% 1,45M 3s\n",
      " 14450K .......... .......... .......... .......... .......... 81% 3,31M 3s\n",
      " 14500K .......... .......... .......... .......... .......... 81% 1,01M 3s\n",
      " 14550K .......... .......... .......... .......... .......... 82% 2,85M 3s\n",
      " 14600K .......... .......... .......... .......... .......... 82% 2,84M 3s\n",
      " 14650K .......... .......... .......... .......... .......... 82% 1,02M 3s\n",
      " 14700K .......... .......... .......... .......... .......... 82% 1,19M 3s\n",
      " 14750K .......... .......... .......... .......... .......... 83%  725K 3s\n",
      " 14800K .......... .......... .......... .......... .......... 83% 1,73M 3s\n",
      " 14850K .......... .......... .......... .......... .......... 83% 20,4M 3s\n",
      " 14900K .......... .......... .......... .......... .......... 84% 2,34M 3s\n",
      " 14950K .......... .......... .......... .......... .......... 84% 1,00M 3s\n",
      " 15000K .......... .......... .......... .......... .......... 84%  932K 3s\n",
      " 15050K .......... .......... .......... .......... .......... 84% 1,13M 3s\n",
      " 15100K .......... .......... .......... .......... .......... 85% 4,65M 3s\n",
      " 15150K .......... .......... .......... .......... .......... 85% 3,15M 2s\n",
      " 15200K .......... .......... .......... .......... .......... 85%  992K 2s\n",
      " 15250K .......... .......... .......... .......... .......... 86% 5,56M 2s\n",
      " 15300K .......... .......... .......... .......... .......... 86%  557K 2s\n",
      " 15350K .......... .......... .......... .......... .......... 86% 14,0M 2s\n",
      " 15400K .......... .......... .......... .......... .......... 86% 2,06M 2s\n",
      " 15450K .......... .......... .......... .......... .......... 87% 1,42M 2s\n",
      " 15500K .......... .......... .......... .......... .......... 87% 2,43M 2s\n",
      " 15550K .......... .......... .......... .......... .......... 87% 1,25M 2s\n",
      " 15600K .......... .......... .......... .......... .......... 87%  938K 2s\n",
      " 15650K .......... .......... .......... .......... .......... 88% 2,85M 2s\n",
      " 15700K .......... .......... .......... .......... .......... 88% 2,42M 2s\n",
      " 15750K .......... .......... .......... .......... .......... 88% 1,20M 2s\n",
      " 15800K .......... .......... .......... .......... .......... 89% 2,10M 2s\n",
      " 15850K .......... .......... .......... .......... .......... 89%  771K 2s\n",
      " 15900K .......... .......... .......... .......... .......... 89% 1,86M 2s\n",
      " 15950K .......... .......... .......... .......... .......... 89% 1,58M 2s\n",
      " 16000K .......... .......... .......... .......... .......... 90% 2,31M 2s\n",
      " 16050K .......... .......... .......... .......... .......... 90% 3,07M 2s\n",
      " 16100K .......... .......... .......... .......... .......... 90%  688K 2s\n",
      " 16150K .......... .......... .......... .......... .......... 91% 2,84M 1s\n",
      " 16200K .......... .......... .......... .......... .......... 91% 1,64M 1s\n",
      " 16250K .......... .......... .......... .......... .......... 91%  743K 1s\n",
      " 16300K .......... .......... .......... .......... .......... 91% 3,12M 1s\n",
      " 16350K .......... .......... .......... .......... .......... 92% 1,30M 1s\n",
      " 16400K .......... .......... .......... .......... .......... 92% 1,79M 1s\n",
      " 16450K .......... .......... .......... .......... .......... 92% 1,87M 1s\n",
      " 16500K .......... .......... .......... .......... .......... 93%  551K 1s\n",
      " 16550K .......... .......... .......... .......... .......... 93% 4,89M 1s\n",
      " 16600K .......... .......... .......... .......... .......... 93% 1,91M 1s\n",
      " 16650K .......... .......... .......... .......... .......... 93% 3,08M 1s\n",
      " 16700K .......... .......... .......... .......... .......... 94% 2,40M 1s\n",
      " 16750K .......... .......... .......... .......... .......... 94% 2,96M 1s\n",
      " 16800K .......... .......... .......... .......... .......... 94%  571K 1s\n",
      " 16850K .......... .......... .......... .......... .......... 95% 1,67M 1s\n",
      " 16900K .......... .......... .......... .......... .......... 95% 2,71M 1s\n",
      " 16950K .......... .......... .......... .......... .......... 95% 2,23M 1s\n",
      " 17000K .......... .......... .......... .......... .......... 95% 2,61M 1s\n",
      " 17050K .......... .......... .......... .......... .......... 96%  645K 1s\n",
      " 17100K .......... .......... .......... .......... .......... 96% 1,32M 1s\n",
      " 17150K .......... .......... .......... .......... .......... 96% 2,23M 1s\n",
      " 17200K .......... .......... .......... .......... .......... 96% 1,58M 0s\n",
      " 17250K .......... .......... .......... .......... .......... 97% 5,10M 0s\n",
      " 17300K .......... .......... .......... .......... .......... 97%  769K 0s\n",
      " 17350K .......... .......... .......... .......... .......... 97% 1,61M 0s\n",
      " 17400K .......... .......... .......... .......... .......... 98% 1,61M 0s\n",
      " 17450K .......... .......... .......... .......... .......... 98%  989K 0s\n",
      " 17500K .......... .......... .......... .......... .......... 98% 6,87M 0s\n",
      " 17550K .......... .......... .......... .......... .......... 98% 4,70M 0s\n",
      " 17600K .......... .......... .......... .......... .......... 99%  990K 0s\n",
      " 17650K .......... .......... .......... .......... .......... 99% 1,89M 0s\n",
      " 17700K .......... .......... .......... .......... .......... 99%  764K 0s\n",
      " 17750K .......... .......... .......... ......               100% 1,39M=16s\n",
      "\n",
      "2025-06-18 03:35:06 (1,07 MB/s) - ‘hmmer-3.3.2.tar.gz’ saved [18213049/18213049]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "wget http://eddylab.org/software/hmmer/hmmer-3.3.2.tar.gz\n",
    "tar xf hmmer-3.3.2.tar.gz\n",
    "cd hmmer-3.3.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "zvgA0W27bgUz"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "FZjhG2leSWW9",
    "outputId": "24a7bc67-0348-453a-ea3c-0edc2f787207"
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd hmmer-3.3.2\n",
    "sudo ./configure\n",
    "sudo make\n",
    "sudo make install"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "Qr4NxI_9JFwA"
   },
   "source": [
    "Скачиваем Pfam\n",
    "По ftp доступны разны версии базы. Нас интересует последняя в формате hmm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "rfK6bdb-Tj3E"
   },
   "outputs": [],
   "source": [
    "#!wget --spider http://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam25.0/Pfam-A.full.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "N-1Z8QTKS-8y"
   },
   "outputs": [],
   "source": [
    "#!wget ftp://ftp.sanger.ac.uk/pub/databases/Pfam/releases/Pfam25.0/Pfam-A.full.gz."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "XfY1wB3oUvL6",
    "outputId": "f2925f8d-a662-413b-d33c-a1524e5c2d3b"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-06-18 03:38:05--  http://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam35.0/Pfam-A.hmm.gz\n",
      "Resolving ftp.ebi.ac.uk (ftp.ebi.ac.uk)... 193.62.193.165\n",
      "Connecting to ftp.ebi.ac.uk (ftp.ebi.ac.uk)|193.62.193.165|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 293000230 (279M) [application/x-gzip]\n",
      "Saving to: ‘Pfam-A.hmm.gz’\n",
      "\n",
      "Pfam-A.hmm.gz       100%[===================>] 279,43M  2,80MB/s    in 2m 34s  \n",
      "\n",
      "2025-06-18 03:40:39 (1,81 MB/s) - ‘Pfam-A.hmm.gz’ saved [293000230/293000230]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget http://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam35.0/Pfam-A.hmm.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "nWnlyTK9Jo0Q"
   },
   "source": [
    "Извлекаем из архива"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "AXN716JmU-O0"
   },
   "outputs": [],
   "source": [
    "!gunzip Pfam-A.hmm.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "CHrwjpTlJtbd"
   },
   "source": [
    "Создаем нужные для hmmer'a файлы"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "abD_KzEBWVqB",
    "outputId": "d31ca453-8428-4f79-8836-286e3b8af1df"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working...    done.\n",
      "Pressed and indexed 19632 HMMs (19632 names and 19632 accessions).\n",
      "Models pressed into binary file:   Pfam-A.hmm.h3m\n",
      "SSI index for binary model file:   Pfam-A.hmm.h3i\n",
      "Profiles (MSV part) pressed into:  Pfam-A.hmm.h3f\n",
      "Profiles (remainder) pressed into: Pfam-A.hmm.h3p\n"
     ]
    }
   ],
   "source": [
    "!hmmpress Pfam-A.hmm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "h5ajvyOxoaS5",
    "outputId": "f1e36d33-5f1b-40b1-8d2d-4d449e4839b4"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HMMER3/f [3.1b2 | February 2015]\n",
      "NAME  1-cysPrx_C\n",
      "ACC   PF10417.12\n",
      "DESC  C-terminal domain of 1-Cys peroxiredoxin\n",
      "LENG  40\n",
      "ALPH  amino\n",
      "RF    no\n",
      "MM    no\n",
      "CONS  yes\n",
      "CS    yes\n",
      "MAP   yes\n",
      "DATE  Thu Nov  4 19:20:02 2021\n",
      "NSEQ  40\n",
      "EFFN  17.426758\n",
      "CKSUM 4086680297\n",
      "GA    21.1 21.1;\n",
      "TC    21.1 21.1;\n",
      "NC    21 21;\n",
      "BM    hmmbuild HMM.ann SEED.ann\n",
      "SM    hmmsearch -Z 61295632 -E 1000 --cpu 4 HMM pfamseq\n",
      "STATS LOCAL MSV       -7.5463  0.71948\n",
      "STATS LOCAL VITERBI   -7.8624  0.71948\n",
      "STATS LOCAL FORWARD   -4.3303  0.71948\n",
      "HMM          A        C        D        E        F        G        H        I        K        L        M        N        P        Q        R        S        T        V        W        Y   \n",
      "            m->m     m->i     m->d     i->m     i->i     d->m     d->d\n",
      "  COMPO   2.28046  4.31208  2.83393  2.63913  3.90855  2.69988  3.89812  3.33401  2.56310  2.85023  3.99954  3.22924  2.52123  2.90328  3.31238  2.94055  2.70512  2.59551  3.49266  3.82715\n",
      "          2.68618  4.42225  2.77519  2.73123  3.46354  2.40513  3.72494  3.29354  2.67741  2.69355  4.24690  2.90347  2.73739  3.18146  2.89801  2.37887  2.77519  2.98518  4.58477  3.61503\n",
      "          0.00226  6.48754  7.20989  0.61958  0.77255  0.00000        *\n",
      "      1   0.29666  6.14436  6.78514  6.79783  7.06332  2.55785  7.22049  6.57837  6.66651  6.27638  3.28757  5.91223  5.83978  6.69238  6.58162  2.20136  4.83343  5.59959  8.41086  7.43107      1 A - - H\n",
      "          2.68618  4.42225  2.77519  2.73123  3.46354  2.40513  3.72494  3.29354  2.67741  2.69355  4.24690  2.90347  2.73739  3.18146  2.89801  2.37887  2.77519  2.98518  4.58477  3.61503\n",
      "          0.00226  6.48754  7.20989  0.61958  0.77255  0.48576  0.95510\n",
      "      2   4.59591  5.92009  6.57211  5.96147  1.92899  5.81035  6.10135  2.33093  5.75927  0.69439  2.86149  5.97820  6.07717  5.78793  5.72916  5.13924  4.81708  2.59612  3.18569  3.35842      2 l - - H\n",
      "          2.68618  4.42225  2.77519  2.73123  3.46354  2.40513  3.72494  3.29354  2.67741  2.69355  4.24690  2.90347  2.73739  3.18146  2.89801  2.37887  2.77519  2.98518  4.58477  3.61503\n",
      "          0.00226  6.48754  7.20989  0.61958  0.77255  0.48576  0.95510\n",
      "      3   4.81290  7.05274  3.71696  4.47757  6.60126  5.41623  3.72993  5.92180  2.06538  3.59487  6.10993  4.89014  5.75663  0.42291  2.54802  4.76779  4.95656  5.56452  7.24472  6.08615      3 Q - - H\n",
      "          2.68618  4.42225  2.77519  2.73123  3.46354  2.40513  3.72494  3.29354  2.67741  2.69355  4.24690  2.90347  2.73739  3.18146  2.89801  2.37887  2.77519  2.98518  4.58477  3.61503\n",
      "          0.00226  6.48754  7.20989  0.61958  0.77255  0.48576  0.95510\n",
      "      4   2.64100  5.28125  5.84007  3.33362  1.80025  5.06896  2.72827  3.71332  5.01717  1.75203  2.65498  5.22280  5.43290  5.15329  5.03455  4.37913  1.80041  2.31249  5.90246  2.63298      4 l - - H\n",
      "          2.68618  4.42225  2.77519  2.73123  3.46354  2.40513  3.72494  3.29354  2.67741  2.69355  4.24690  2.90347  2.73739  3.18146  2.89801  2.37887  2.77519  2.98518  4.58477  3.61503\n",
      "          0.00226  6.48754  7.20989  0.61958  0.77255  0.48576  0.95510\n"
     ]
    }
   ],
   "source": [
    "!head -n 40 Pfam-A.hmm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "zL3sy2rvJ5Wl"
   },
   "source": [
    "Смотрим статистику"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "_5JS0hV7Z-Zq",
    "outputId": "0d58afe3-814c-43db-8559-e5eaec807c88"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmstat :: display summary statistics for a profile file\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "#\n",
      "# idx  name                 accession        nseq eff_nseq      M relent   info p relE compKL\n",
      "# ---- -------------------- ------------ -------- -------- ------ ------ ------ ------ ------\n",
      "1      1-cysPrx_C           PF10417.12         40    17.43     40   1.37   1.32   1.29   0.09\n",
      "2      120_Rick_ant         PF12574.11          2     0.45    238   0.59   0.60   0.52   0.02\n",
      "3      12TM_1               PF09847.12          7     2.28    449   0.59   0.62   0.51   0.09\n",
      "4      14-3-3               PF00244.23        149     2.50    222   0.59   0.59   0.52   0.05\n",
      "5      17kDa_Anti_2         PF16998.8           6     0.99    116   0.59   0.57   0.52   0.03\n",
      "6      2-Hacid_dh           PF00389.33         78     5.69    134   0.59   0.62   0.44   0.03\n",
      "7      2-Hacid_dh_C         PF02826.22        135     3.55    178   0.59   0.60   0.53   0.01\n",
      "8      2-oxoacid_dh         PF00198.26         70     3.33    233   0.59   0.61   0.53   0.02\n",
      "9      2-oxogl_dehyd_N      PF16078.8         208    11.82     41   1.34   1.23   1.33   0.13\n",
      "10     2-ph_phosp           PF04029.17        125     3.02    226   0.59   0.62   0.51   0.02\n",
      "11     2-thiour_desulf      PF04463.15        338     2.11    137   0.59   0.57   0.46   0.02\n",
      "12     23ISL                PF16620.8          16     1.48    162   0.59   0.61   0.54   0.07\n"
     ]
    }
   ],
   "source": [
    "!hmmstat Pfam-A.hmm.h3m | head -n 20"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "xBqiKE1AdaCY"
   },
   "source": [
    "The columns are:\n",
    "\n",
    "**idx**\n",
    "The index of this profile, numbering each on in the file starting from 1.\n",
    "\n",
    "**name**\n",
    "The name of the profile.\n",
    "\n",
    "**accession**\n",
    "The optional accession of the profile, or \"-\" if there is none.\n",
    "\n",
    "**nseq**\n",
    "The number of sequences that the profile was estimated from.\n",
    "\n",
    "**eff_nseq**\n",
    "The effective number of sequences that the profile was estimated from, after HMMER applied an effective sequence number calculation such as the default entropy weighting.\n",
    "\n",
    "**M**\n",
    "The length of the model in consensus residues (match states).\n",
    "\n",
    "**relent**\n",
    "Mean relative entropy per match state, in bits. This is the expected (mean) score per consensus position. This is what the default entropy-weighting method for effective sequence number estimation focuses on, so for default HMMER3 models, you expect this value to reflect the default target for entropy-weighting.\n",
    "\n",
    "**info**\n",
    "Mean information content per match state, in bits. Probably not useful. Information content is a slightly different calculation than relative entropy.\n",
    "\n",
    "**p relE**\n",
    "Mean positional relative entropy, in bits. This is a fancier version of the per-match-state relative entropy, taking into account the transition (insertion/deletion) probabilities; it may be a more accurate estimation of the average score contributed per model consensus position.\n",
    "\n",
    "**compKL**\n",
    "Kullback-Leibler distance between the model's overall average residue composition and the default background frequency distribution. The higher this number, the more biased the residue composition of the profile is. Highly biased profiles can slow the HMMER3 acceleration pipeline, by causing too many nonhomologous sequences to pass the filters."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "2N22F0eaKD4E"
   },
   "source": [
    "Индексируем файл"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "2CZAZX4yU-Tw",
    "outputId": "ea9a2ee2-b839-4c31-bdfd-3ad6bdec1cac"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working...    done.\n",
      "Indexed 19632 HMMs (19632 names and 19632 accessions).\n",
      "SSI index written to file Pfam-A.hmm.h3m.ssi\n"
     ]
    }
   ],
   "source": [
    "!hmmfetch --index Pfam-A.hmm.h3m"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "fnP4zCPwFKe4"
   },
   "source": [
    "Скачиваем протеом"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "0rB1i34WN9Sa",
    "outputId": "82287bb1-7edc-44b2-c90e-9d1754dddf4d"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-06-18 03:45:47--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/902/829/305/GCA_902829305.4_PGI_SPEXI_v6/GCA_902829305.4_PGI_SPEXI_v6_protein.faa.gz\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.11, 130.14.250.31, 130.14.250.10, ...\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.11|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 4096709 (3,9M) [application/x-gzip]\n",
      "Saving to: ‘GCA_902829305.4_PGI_SPEXI_v6_protein.faa.gz’\n",
      "\n",
      "GCA_902829305.4_PGI 100%[===================>]   3,91M   647KB/s    in 8,4s    \n",
      "\n",
      "2025-06-18 03:45:58 (478 KB/s) - ‘GCA_902829305.4_PGI_SPEXI_v6_protein.faa.gz’ saved [4096709/4096709]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/902/829/305/GCA_902829305.4_PGI_SPEXI_v6/GCA_902829305.4_PGI_SPEXI_v6_protein.faa.gz\n",
    "!mv GCA_902829305.4_PGI_SPEXI_v6_protein.faa.gz protein.faa.gz\n",
    "!gunzip  protein.faa.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "c6vDzm96arV_"
   },
   "source": [
    "Для получения названий генов скачаем и подготовим аннотацию"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "idsNP8PXU_9J",
    "outputId": "6befde1f-76ce-4942-a385-3897563dba47"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-06-18 03:46:46--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/902/829/305/GCA_902829305.4_PGI_SPEXI_v6/GCA_902829305.4_PGI_SPEXI_v6_genomic.gff.gz\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.12, 130.14.250.10, 130.14.250.31, ...\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.12|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 4243215 (4,0M) [application/x-gzip]\n",
      "Saving to: ‘GCA_902829305.4_PGI_SPEXI_v6_genomic.gff.gz’\n",
      "\n",
      "GCA_902829305.4_PGI 100%[===================>]   4,05M   434KB/s    in 16s     \n",
      "\n",
      "2025-06-18 03:47:04 (257 KB/s) - ‘GCA_902829305.4_PGI_SPEXI_v6_genomic.gff.gz’ saved [4243215/4243215]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/902/829/305/GCA_902829305.4_PGI_SPEXI_v6/GCA_902829305.4_PGI_SPEXI_v6_genomic.gff.gz\n",
    "!mv GCA_902829305.4_PGI_SPEXI_v6_genomic.gff.gz genomic.gff.gz\n",
    "!gunzip genomic.gff.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 424
    },
    "id": "S_S4G6diS2vM",
    "outputId": "05de7135-f8d9-42b7-e0ec-7f514a409daa"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LR824602.2</td>\n",
       "      <td>EMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>433449.0</td>\n",
       "      <td>435904.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene-SPEXI_LOCUS1;Name=SPEXI_LOCUS1;gbkey=G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LR824602.2</td>\n",
       "      <td>EMBL</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>433449.0</td>\n",
       "      <td>435904.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna-SPEXI_LOCUS1;Parent=gene-SPEXI_LOCUS1;N...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LR824602.2</td>\n",
       "      <td>EMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>433449.0</td>\n",
       "      <td>433533.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon-SPEXI_LOCUS1-1;Parent=rna-SPEXI_LOCUS1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LR824602.2</td>\n",
       "      <td>EMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>433946.0</td>\n",
       "      <td>434073.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon-SPEXI_LOCUS1-2;Parent=rna-SPEXI_LOCUS1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LR824602.2</td>\n",
       "      <td>EMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>434612.0</td>\n",
       "      <td>434836.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon-SPEXI_LOCUS1-3;Parent=rna-SPEXI_LOCUS1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365096</th>\n",
       "      <td>CADEPN050000003.1</td>\n",
       "      <td>EMBL</td>\n",
       "      <td>region</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1850385.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CADEPN050000003.1:1..1850385;Dbxref=taxon:7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365097</th>\n",
       "      <td>##sequence-region CADEPN050000004.1 1 706148</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365098</th>\n",
       "      <td>##species https://www.ncbi.nlm.nih.gov/Taxonom...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365099</th>\n",
       "      <td>CADEPN050000004.1</td>\n",
       "      <td>EMBL</td>\n",
       "      <td>region</td>\n",
       "      <td>1.0</td>\n",
       "      <td>706148.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=CADEPN050000004.1:1..706148;Dbxref=taxon:71...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365100</th>\n",
       "      <td>###</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>365101 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        0     1       2  \\\n",
       "0                                              LR824602.2  EMBL    gene   \n",
       "1                                              LR824602.2  EMBL    mRNA   \n",
       "2                                              LR824602.2  EMBL    exon   \n",
       "3                                              LR824602.2  EMBL    exon   \n",
       "4                                              LR824602.2  EMBL    exon   \n",
       "...                                                   ...   ...     ...   \n",
       "365096                                  CADEPN050000003.1  EMBL  region   \n",
       "365097       ##sequence-region CADEPN050000004.1 1 706148   NaN     NaN   \n",
       "365098  ##species https://www.ncbi.nlm.nih.gov/Taxonom...   NaN     NaN   \n",
       "365099                                  CADEPN050000004.1  EMBL  region   \n",
       "365100                                                ###   NaN     NaN   \n",
       "\n",
       "               3          4    5    6    7  \\\n",
       "0       433449.0   435904.0    .    +    .   \n",
       "1       433449.0   435904.0    .    +    .   \n",
       "2       433449.0   433533.0    .    +    .   \n",
       "3       433946.0   434073.0    .    +    .   \n",
       "4       434612.0   434836.0    .    +    .   \n",
       "...          ...        ...  ...  ...  ...   \n",
       "365096       1.0  1850385.0    .    +    .   \n",
       "365097       NaN        NaN  NaN  NaN  NaN   \n",
       "365098       NaN        NaN  NaN  NaN  NaN   \n",
       "365099       1.0   706148.0    .    +    .   \n",
       "365100       NaN        NaN  NaN  NaN  NaN   \n",
       "\n",
       "                                                        8  \n",
       "0       ID=gene-SPEXI_LOCUS1;Name=SPEXI_LOCUS1;gbkey=G...  \n",
       "1       ID=rna-SPEXI_LOCUS1;Parent=gene-SPEXI_LOCUS1;N...  \n",
       "2       ID=exon-SPEXI_LOCUS1-1;Parent=rna-SPEXI_LOCUS1...  \n",
       "3       ID=exon-SPEXI_LOCUS1-2;Parent=rna-SPEXI_LOCUS1...  \n",
       "4       ID=exon-SPEXI_LOCUS1-3;Parent=rna-SPEXI_LOCUS1...  \n",
       "...                                                   ...  \n",
       "365096  ID=CADEPN050000003.1:1..1850385;Dbxref=taxon:7...  \n",
       "365097                                                NaN  \n",
       "365098                                                NaN  \n",
       "365099  ID=CADEPN050000004.1:1..706148;Dbxref=taxon:71...  \n",
       "365100                                                NaN  \n",
       "\n",
       "[365101 rows x 9 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "genomic = pd.read_csv('genomic.gff', skiprows=8, header=None, sep='\\t')\n",
    "genomic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "id": "2c2mJBvJY3cR"
   },
   "outputs": [],
   "source": [
    "from tqdm.auto import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGNC approved symbol</th>\n",
       "      <th>Product</th>\n",
       "      <th>Function</th>\n",
       "      <th>Protein complex</th>\n",
       "      <th>Target entity</th>\n",
       "      <th>HGNC approved name</th>\n",
       "      <th>UniProt ID (human)</th>\n",
       "      <th>Pfam domains</th>\n",
       "      <th>UniProt ID (mouse)</th>\n",
       "      <th>Modification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>U</td>\n",
       "      <td>RNA modification</td>\n",
       "      <td>APOB_mRNA_editosome</td>\n",
       "      <td>mRNA, mC</td>\n",
       "      <td>APOBEC1 complementation factor</td>\n",
       "      <td>A1CF_HUMAN</td>\n",
       "      <td>DND1_DSRM PF14709 445-523, RRM_1 PF00076 58-12...</td>\n",
       "      <td>A1CF_MOUSE</td>\n",
       "      <td>RNA deamination</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACINU</td>\n",
       "      <td>#</td>\n",
       "      <td>RNA modification</td>\n",
       "      <td>#</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>Apoptotic chromatin condensation inducer in th...</td>\n",
       "      <td>ACINU_HUMAN</td>\n",
       "      <td>PF16294;PF02037</td>\n",
       "      <td>ACINU_MOUSE</td>\n",
       "      <td>Alternative splicing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>#</td>\n",
       "      <td>Chromatin remodeling cofactor</td>\n",
       "      <td>BAF, nBAF, npBAF, PBAF, SWI/SNF-like EPAFB, bB...</td>\n",
       "      <td>#</td>\n",
       "      <td>actin, beta</td>\n",
       "      <td>ACTB_HUMAN</td>\n",
       "      <td>Actin PF00022 2-375</td>\n",
       "      <td>ACTB_MOUSE</td>\n",
       "      <td>#</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACTL6A</td>\n",
       "      <td>#</td>\n",
       "      <td>Chromatin remodeling cofactor</td>\n",
       "      <td>BAF, npBAF, PBAF, SWI/SNF_Brg1(I), SWI/SNF_Brg...</td>\n",
       "      <td>#</td>\n",
       "      <td>actin-like 6A</td>\n",
       "      <td>ACL6A_HUMAN</td>\n",
       "      <td>Actin PF00022 8-429</td>\n",
       "      <td>ACL6A_MOUSE</td>\n",
       "      <td>#</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACTL6B</td>\n",
       "      <td>#</td>\n",
       "      <td>Chromatin remodeling cofactor</td>\n",
       "      <td>BAF, nBAF, PBAF, SWI/SNF_Brg1(I), SWI/SNF_Brg1...</td>\n",
       "      <td>#</td>\n",
       "      <td>actin-like 6B</td>\n",
       "      <td>ACL6B_HUMAN</td>\n",
       "      <td>Actin PF00022 8-426</td>\n",
       "      <td>ACL6B_MOUSE</td>\n",
       "      <td>#</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>774</th>\n",
       "      <td>ZNF687</td>\n",
       "      <td>#</td>\n",
       "      <td>Histone modification erase cofactor</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>zinc finger protein 687</td>\n",
       "      <td>ZN687_HUMAN</td>\n",
       "      <td>zf-C2H2 PF00096 1200-1222, zf-C2H2_4 PF13894 8...</td>\n",
       "      <td>ZN687_MOUSE</td>\n",
       "      <td>Histone acetylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>775</th>\n",
       "      <td>ZNF711</td>\n",
       "      <td>#</td>\n",
       "      <td>Histone modification erase cofactor</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>zinc finger protein 711</td>\n",
       "      <td>ZN711_HUMAN</td>\n",
       "      <td>Pfam-B_16491 PB016491 1-63, Zfx_Zfy_act PF0470...</td>\n",
       "      <td>ZN711_MOUSE</td>\n",
       "      <td>Histone acetylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>776</th>\n",
       "      <td>ZNHIT1</td>\n",
       "      <td>#</td>\n",
       "      <td>Chromatin remodeling cofactor, Histone modific...</td>\n",
       "      <td>SRCAP</td>\n",
       "      <td>#</td>\n",
       "      <td>zinc finger, HIT-type containing 1</td>\n",
       "      <td>ZNHI1_HUMAN</td>\n",
       "      <td>zf-HIT PF04438 113-142</td>\n",
       "      <td>ZNHI1_MOUSE</td>\n",
       "      <td>Histone acetylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777</th>\n",
       "      <td>ZRANB3</td>\n",
       "      <td>#</td>\n",
       "      <td>Chromatin remodeling, Histone modification rea...</td>\n",
       "      <td>#</td>\n",
       "      <td>#</td>\n",
       "      <td>zinc finger, RAN-binding domain containing 3</td>\n",
       "      <td>ZRAB3_HUMAN</td>\n",
       "      <td>HNH PF01844 1009-1051, Helicase_C PF00271 357-...</td>\n",
       "      <td>ZRAB3_MOUSE</td>\n",
       "      <td>Histone methylation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>778</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>#</td>\n",
       "      <td>Histone modification read</td>\n",
       "      <td>ATAC</td>\n",
       "      <td>#</td>\n",
       "      <td>zinc finger, ZZ-type containing 3</td>\n",
       "      <td>ZZZ3_HUMAN</td>\n",
       "      <td>Myb_DNA-binding PF00249 652-703, Pfam-B_19065 ...</td>\n",
       "      <td>ZZZ3_MOUSE</td>\n",
       "      <td>#</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>779 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    HGNC approved symbol Product  \\\n",
       "0                   A1CF       U   \n",
       "1                  ACINU       #   \n",
       "2                   ACTB       #   \n",
       "3                 ACTL6A       #   \n",
       "4                 ACTL6B       #   \n",
       "..                   ...     ...   \n",
       "774               ZNF687       #   \n",
       "775               ZNF711       #   \n",
       "776               ZNHIT1       #   \n",
       "777               ZRANB3       #   \n",
       "778                 ZZZ3       #   \n",
       "\n",
       "                                              Function  \\\n",
       "0                                     RNA modification   \n",
       "1                                     RNA modification   \n",
       "2                        Chromatin remodeling cofactor   \n",
       "3                        Chromatin remodeling cofactor   \n",
       "4                        Chromatin remodeling cofactor   \n",
       "..                                                 ...   \n",
       "774                Histone modification erase cofactor   \n",
       "775                Histone modification erase cofactor   \n",
       "776  Chromatin remodeling cofactor, Histone modific...   \n",
       "777  Chromatin remodeling, Histone modification rea...   \n",
       "778                          Histone modification read   \n",
       "\n",
       "                                       Protein complex Target entity  \\\n",
       "0                                  APOB_mRNA_editosome      mRNA, mC   \n",
       "1                                                    #          mRNA   \n",
       "2    BAF, nBAF, npBAF, PBAF, SWI/SNF-like EPAFB, bB...             #   \n",
       "3    BAF, npBAF, PBAF, SWI/SNF_Brg1(I), SWI/SNF_Brg...             #   \n",
       "4    BAF, nBAF, PBAF, SWI/SNF_Brg1(I), SWI/SNF_Brg1...             #   \n",
       "..                                                 ...           ...   \n",
       "774                                                  #             #   \n",
       "775                                                  #             #   \n",
       "776                                              SRCAP             #   \n",
       "777                                                  #             #   \n",
       "778                                               ATAC             #   \n",
       "\n",
       "                                    HGNC approved name UniProt ID (human)  \\\n",
       "0                       APOBEC1 complementation factor         A1CF_HUMAN   \n",
       "1    Apoptotic chromatin condensation inducer in th...        ACINU_HUMAN   \n",
       "2                                          actin, beta         ACTB_HUMAN   \n",
       "3                                        actin-like 6A        ACL6A_HUMAN   \n",
       "4                                        actin-like 6B        ACL6B_HUMAN   \n",
       "..                                                 ...                ...   \n",
       "774                            zinc finger protein 687        ZN687_HUMAN   \n",
       "775                            zinc finger protein 711        ZN711_HUMAN   \n",
       "776                 zinc finger, HIT-type containing 1        ZNHI1_HUMAN   \n",
       "777       zinc finger, RAN-binding domain containing 3        ZRAB3_HUMAN   \n",
       "778                  zinc finger, ZZ-type containing 3         ZZZ3_HUMAN   \n",
       "\n",
       "                                          Pfam domains UniProt ID (mouse)  \\\n",
       "0    DND1_DSRM PF14709 445-523, RRM_1 PF00076 58-12...         A1CF_MOUSE   \n",
       "1                                      PF16294;PF02037        ACINU_MOUSE   \n",
       "2                                  Actin PF00022 2-375         ACTB_MOUSE   \n",
       "3                                  Actin PF00022 8-429        ACL6A_MOUSE   \n",
       "4                                  Actin PF00022 8-426        ACL6B_MOUSE   \n",
       "..                                                 ...                ...   \n",
       "774  zf-C2H2 PF00096 1200-1222, zf-C2H2_4 PF13894 8...        ZN687_MOUSE   \n",
       "775  Pfam-B_16491 PB016491 1-63, Zfx_Zfy_act PF0470...        ZN711_MOUSE   \n",
       "776                             zf-HIT PF04438 113-142        ZNHI1_MOUSE   \n",
       "777  HNH PF01844 1009-1051, Helicase_C PF00271 357-...        ZRAB3_MOUSE   \n",
       "778  Myb_DNA-binding PF00249 652-703, Pfam-B_19065 ...         ZZZ3_MOUSE   \n",
       "\n",
       "             Modification  \n",
       "0         RNA deamination  \n",
       "1    Alternative splicing  \n",
       "2                       #  \n",
       "3                       #  \n",
       "4                       #  \n",
       "..                    ...  \n",
       "774   Histone acetylation  \n",
       "775   Histone acetylation  \n",
       "776   Histone acetylation  \n",
       "777   Histone methylation  \n",
       "778                     #  \n",
       "\n",
       "[779 rows x 10 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(\"pfam_selected.csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "id": "UgpQzhVPWTqE"
   },
   "outputs": [],
   "source": [
    "import re\n",
    "domains = set()\n",
    "for el in df[\"Pfam domains\"]:\n",
    "    for domain in re.split('[^a-zA-Z0-9]', el):\n",
    "        if domain.startswith(\"PF\"):\n",
    "            domains.add(domain)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "!hmmstat Pfam-A.hmm.h3m > hmmstat.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19632"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "accession_to_version = dict()\n",
    "with open(\"hmmstat.txt\") as f:\n",
    "    for line in f:\n",
    "        if line[0].isdigit():\n",
    "            accession_with_version = line.split()[2]\n",
    "            accession, version = accession_with_version.split('.')\n",
    "            accession_to_version[accession] = accession_to_version.get(accession, []) + [version]\n",
    "len(accession_to_version)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    510     510    5527\n",
      "PF05891.15\n",
      "PF00583.28\n",
      "PF00271.34\n",
      "PF01388.24\n",
      "PF04762.15\n",
      "PF10426.12\n",
      "PF01344.28\n",
      "PF12253.11\n",
      "PF00888.25\n",
      "PF08169.14\n"
     ]
    }
   ],
   "source": [
    "with open(\"keys.txt\", \"w\") as f:\n",
    "    for domain in domains:\n",
    "        for version in accession_to_version.get(domain, []):\n",
    "            print(f\"{domain}.{version}\", file=f)\n",
    "!cat keys.txt | wc\n",
    "!head keys.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "!hmmfetch -f Pfam-A.hmm.h3m keys.txt > hmmfetch.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "!hmmsearch hmmfetch.txt protein.faa > hmmsearch.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmsearch :: search profile(s) against a sequence database\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# query HMM file:                  hmmfetch.txt\n",
      "# target sequence database:        protein.faa\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "Query:       Methyltransf_PK  [M=218]\n",
      "Accession:   PF05891.15\n",
      "Description: AdoMet dependent proline di-methyltransferase\n",
      "Scores for complete sequences (score includes all domains):\n",
      "   --- full sequence ---   --- best 1 domain ---    -#dom-\n",
      "    E-value  score  bias    E-value  score  bias    exp  N  Sequence     Description\n",
      "    ------- ------ -----    ------- ------ -----   ---- --  --------     -----------\n",
      "    1.9e-76  256.0   0.0    2.4e-76  255.7   0.0    1.0  1  CAH0700664.1  unnamed protein product [Spodoptera exigua]\n",
      "    1.9e-75  252.8   0.0    2.2e-75  252.6   0.0    1.0  1  CAH0691802.1  unnamed protein product [Spodoptera exigua]\n",
      "      0.006   15.8   0.1      0.016   14.4   0.0    1.6  2  CAH0699521.1  unnamed protein product [Spodoptera exigua]\n",
      "\n",
      "\n",
      "Domain annotation for each sequence (and alignments):\n",
      ">> CAH0700664.1  unnamed protein product [Spodoptera exigua]\n",
      "   #    score  bias  c-Evalue  i-Evalue hmmfrom  hmm to    alifrom  ali to    envfrom  env to     acc\n",
      " ---   ------ ----- --------- --------- ------- -------    ------- -------    ------- -------    ----\n",
      "   1 !  255.7   0.0   5.1e-80   2.4e-76       4     217 ..      10     220 ..       8     221 .. 0.97\n",
      "\n",
      "  Alignments for each domain:\n",
      "  == domain 1  score: 255.7 bits;  conditional E-value: 5.1e-80\n",
      "  Methyltransf_PK   4 evyekaiaYWesveadvdGvLGGygavskiDvkgsrkflkklkkeslpgkrevkvaldcGaGiGrvtknlLlkllskvDlvepvekflekakee 97 \n",
      "                      + y+k+++YW ++ ++vdGvLGG+g +s++D++gs+ flk+l++ +     ++kvaldcGaGiGr+tk lLl++++ vD+vep ekf++ +ke+\n",
      "     CAH0700664.1  10 INYKKSLQYWADIPPTVDGVLGGFGFISNADIEGSELFLKSLLSSES--GPSTKVALDCGAGIGRITKYLLLPYFESVDVVEPDEKFINSIKEY 101\n",
      "                      679****************************************9954..67899**************************************** PP\n",
      "\n",
      "  Methyltransf_PK  98 laeekkkvgkvyvvglqeftPeegrYdliWiqWcvghLtDedlvaflkrckkalkpnGlivvkenvakeekdelDkeDssvtrseaklkklfkk 191\n",
      "                      ++e   k g++y vglqef+P  ++Yd+iW+qW++g LtDedl a+l rc  +l  nG+ivvkenv++++k e D +Dssvtr+ +++ ++fkk\n",
      "     CAH0700664.1 102 VGELAPKLGSLYKVGLQEFVPV-RKYDVIWNQWVLGYLTDEDLEAYLIRCRDSLSGNGIIVVKENVTSSGKVEKDSTDSSVTRTLKQYVNIFKK 194\n",
      "                      *********************6.79********************************************************************* PP\n",
      "\n",
      "  Methyltransf_PK 192 agLklvaeelqkglPkelypvkmyaL 217\n",
      "                      ++Lk +++ +q ++P+ +ypv+m+aL\n",
      "     CAH0700664.1 195 CNLKRIKQCKQANFPNGIYPVYMFAL 220\n",
      "                      *************************9 PP\n",
      "\n",
      ">> CAH0691802.1  unnamed protein product [Spodoptera exigua]\n",
      "   #    score  bias  c-Evalue  i-Evalue hmmfrom  hmm to    alifrom  ali to    envfrom  env to     acc\n",
      " ---   ------ ----- --------- --------- ------- -------    ------- -------    ------- -------    ----\n",
      "   1 !  252.6   0.0   4.6e-79   2.2e-75       2     218 .]       3     216 ..       2     216 .. 0.97\n",
      "\n",
      "  Alignments for each domain:\n"
     ]
    }
   ],
   "source": [
    "!head -n 50 hmmsearch.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "510"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "queries = open(\"hmmsearch.txt\").read().split(\"\\nQuery:\")[1:]\n",
    "len(queries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-06-18 20:12:28--  https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/902/829/305/GCA_902829305.4_PGI_SPEXI_v6/GCA_902829305.4_PGI_SPEXI_v6_feature_table.txt.gz\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.13, 130.14.250.31, 130.14.250.7, ...\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.13|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 609518 (595K) [application/x-gzip]\n",
      "Saving to: ‘GCA_902829305.4_PGI_SPEXI_v6_feature_table.txt.gz’\n",
      "\n",
      "GCA_902829305.4_PGI 100%[===================>] 595,23K   271KB/s    in 2,2s    \n",
      "\n",
      "2025-06-18 20:12:32 (271 KB/s) - ‘GCA_902829305.4_PGI_SPEXI_v6_feature_table.txt.gz’ saved [609518/609518]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/902/829/305/GCA_902829305.4_PGI_SPEXI_v6/GCA_902829305.4_PGI_SPEXI_v6_feature_table.txt.gz\n",
    "!mv GCA_902829305.4_PGI_SPEXI_v6_feature_table.txt.gz feature_table.txt.gz\n",
    "!gunzip feature_table.txt.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_137760/1794484129.py:1: DtypeWarning: Columns (5) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  feature_table = pd.read_csv(\"feature_table.txt\", sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># feature</th>\n",
       "      <th>class</th>\n",
       "      <th>assembly</th>\n",
       "      <th>assembly_unit</th>\n",
       "      <th>seq_type</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>genomic_accession</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>product_accession</th>\n",
       "      <th>non-redundant_refseq</th>\n",
       "      <th>related_accession</th>\n",
       "      <th>name</th>\n",
       "      <th>symbol</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>feature_interval_length</th>\n",
       "      <th>product_length</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GCA_902829305.4</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1</td>\n",
       "      <td>LR824602.2</td>\n",
       "      <td>433449</td>\n",
       "      <td>435904</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPEXI_LOCUS1</td>\n",
       "      <td>2456</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_902829305.4</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1</td>\n",
       "      <td>LR824602.2</td>\n",
       "      <td>433449</td>\n",
       "      <td>435904</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAD0243785.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPEXI_LOCUS1</td>\n",
       "      <td>905</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDS</td>\n",
       "      <td>with_protein</td>\n",
       "      <td>GCA_902829305.4</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1</td>\n",
       "      <td>LR824602.2</td>\n",
       "      <td>433449</td>\n",
       "      <td>435710</td>\n",
       "      <td>+</td>\n",
       "      <td>CAD0243785.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPEXI_LOCUS1</td>\n",
       "      <td>711</td>\n",
       "      <td>236.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GCA_902829305.4</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1</td>\n",
       "      <td>LR824602.2</td>\n",
       "      <td>4897126</td>\n",
       "      <td>4901728</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPEXI_LOCUS2</td>\n",
       "      <td>4603</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_902829305.4</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1</td>\n",
       "      <td>LR824602.2</td>\n",
       "      <td>4897126</td>\n",
       "      <td>4901728</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAH0663572.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPEXI_LOCUS2</td>\n",
       "      <td>2402</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41024</th>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_902829305.4</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Z</td>\n",
       "      <td>LR824601.2</td>\n",
       "      <td>20021997</td>\n",
       "      <td>20026454</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAH0703770.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPEXI_LOCUS13001</td>\n",
       "      <td>1646</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41025</th>\n",
       "      <td>CDS</td>\n",
       "      <td>with_protein</td>\n",
       "      <td>GCA_902829305.4</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Z</td>\n",
       "      <td>LR824601.2</td>\n",
       "      <td>20021997</td>\n",
       "      <td>20025297</td>\n",
       "      <td>+</td>\n",
       "      <td>CAH0703770.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPEXI_LOCUS13001</td>\n",
       "      <td>489</td>\n",
       "      <td>162.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41026</th>\n",
       "      <td>gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>GCA_902829305.4</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Z</td>\n",
       "      <td>LR824601.2</td>\n",
       "      <td>20080804</td>\n",
       "      <td>20082992</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPEXI_LOCUS13002</td>\n",
       "      <td>2189</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41027</th>\n",
       "      <td>mRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCA_902829305.4</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Z</td>\n",
       "      <td>LR824601.2</td>\n",
       "      <td>20080804</td>\n",
       "      <td>20082992</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAH0703771.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPEXI_LOCUS13002</td>\n",
       "      <td>840</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41028</th>\n",
       "      <td>CDS</td>\n",
       "      <td>with_protein</td>\n",
       "      <td>GCA_902829305.4</td>\n",
       "      <td>Primary Assembly</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Z</td>\n",
       "      <td>LR824601.2</td>\n",
       "      <td>20080804</td>\n",
       "      <td>20082992</td>\n",
       "      <td>-</td>\n",
       "      <td>CAH0703771.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPEXI_LOCUS13002</td>\n",
       "      <td>840</td>\n",
       "      <td>279.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41029 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      # feature           class         assembly     assembly_unit  \\\n",
       "0          gene  protein_coding  GCA_902829305.4  Primary Assembly   \n",
       "1          mRNA             NaN  GCA_902829305.4  Primary Assembly   \n",
       "2           CDS    with_protein  GCA_902829305.4  Primary Assembly   \n",
       "3          gene  protein_coding  GCA_902829305.4  Primary Assembly   \n",
       "4          mRNA             NaN  GCA_902829305.4  Primary Assembly   \n",
       "...         ...             ...              ...               ...   \n",
       "41024      mRNA             NaN  GCA_902829305.4  Primary Assembly   \n",
       "41025       CDS    with_protein  GCA_902829305.4  Primary Assembly   \n",
       "41026      gene  protein_coding  GCA_902829305.4  Primary Assembly   \n",
       "41027      mRNA             NaN  GCA_902829305.4  Primary Assembly   \n",
       "41028       CDS    with_protein  GCA_902829305.4  Primary Assembly   \n",
       "\n",
       "         seq_type chromosome genomic_accession     start       end strand  \\\n",
       "0      chromosome          1        LR824602.2    433449    435904      +   \n",
       "1      chromosome          1        LR824602.2    433449    435904      +   \n",
       "2      chromosome          1        LR824602.2    433449    435710      +   \n",
       "3      chromosome          1        LR824602.2   4897126   4901728      -   \n",
       "4      chromosome          1        LR824602.2   4897126   4901728      -   \n",
       "...           ...        ...               ...       ...       ...    ...   \n",
       "41024  chromosome          Z        LR824601.2  20021997  20026454      +   \n",
       "41025  chromosome          Z        LR824601.2  20021997  20025297      +   \n",
       "41026  chromosome          Z        LR824601.2  20080804  20082992      -   \n",
       "41027  chromosome          Z        LR824601.2  20080804  20082992      -   \n",
       "41028  chromosome          Z        LR824601.2  20080804  20082992      -   \n",
       "\n",
       "      product_accession  non-redundant_refseq related_accession  name  symbol  \\\n",
       "0                   NaN                   NaN               NaN   NaN     NaN   \n",
       "1                   NaN                   NaN      CAD0243785.1   NaN     NaN   \n",
       "2          CAD0243785.1                   NaN               NaN   NaN     NaN   \n",
       "3                   NaN                   NaN               NaN   NaN     NaN   \n",
       "4                   NaN                   NaN      CAH0663572.1   NaN     NaN   \n",
       "...                 ...                   ...               ...   ...     ...   \n",
       "41024               NaN                   NaN      CAH0703770.1   NaN     NaN   \n",
       "41025      CAH0703770.1                   NaN               NaN   NaN     NaN   \n",
       "41026               NaN                   NaN               NaN   NaN     NaN   \n",
       "41027               NaN                   NaN      CAH0703771.1   NaN     NaN   \n",
       "41028      CAH0703771.1                   NaN               NaN   NaN     NaN   \n",
       "\n",
       "       GeneID         locus_tag  feature_interval_length  product_length  \\\n",
       "0         NaN      SPEXI_LOCUS1                     2456             NaN   \n",
       "1         NaN      SPEXI_LOCUS1                      905             NaN   \n",
       "2         NaN      SPEXI_LOCUS1                      711           236.0   \n",
       "3         NaN      SPEXI_LOCUS2                     4603             NaN   \n",
       "4         NaN      SPEXI_LOCUS2                     2402             NaN   \n",
       "...       ...               ...                      ...             ...   \n",
       "41024     NaN  SPEXI_LOCUS13001                     1646             NaN   \n",
       "41025     NaN  SPEXI_LOCUS13001                      489           162.0   \n",
       "41026     NaN  SPEXI_LOCUS13002                     2189             NaN   \n",
       "41027     NaN  SPEXI_LOCUS13002                      840             NaN   \n",
       "41028     NaN  SPEXI_LOCUS13002                      840           279.0   \n",
       "\n",
       "      attributes  \n",
       "0            NaN  \n",
       "1            NaN  \n",
       "2            NaN  \n",
       "3            NaN  \n",
       "4            NaN  \n",
       "...          ...  \n",
       "41024        NaN  \n",
       "41025        NaN  \n",
       "41026        NaN  \n",
       "41027        NaN  \n",
       "41028        NaN  \n",
       "\n",
       "[41029 rows x 20 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_table = pd.read_csv(\"feature_table.txt\", sep='\\t')\n",
    "feature_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "466\n"
     ]
    }
   ],
   "source": [
    "domain_gene_table = {\n",
    "    \"domain\": [],\n",
    "    \"gene_name\": [],\n",
    "    \"start\": [],\n",
    "    \"end\": [],\n",
    "    \"strand\": []\n",
    "}\n",
    "\n",
    "for query in queries:\n",
    "    if \"[No hits detected that satisfy reporting thresholds]\" in query:\n",
    "        continue\n",
    "    domain = query.split()[0]\n",
    "    gene_name = query.split('\\n>> ')[1].split()[0]\n",
    "    bebra = feature_table[feature_table[\"product_accession\"] == gene_name]\n",
    "    assert len(bebra) == 1\n",
    "    start = bebra[\"start\"].iloc[0]\n",
    "    end = bebra[\"end\"].iloc[0]\n",
    "    strand = bebra[\"strand\"].iloc[0]\n",
    "\n",
    "    domain_gene_table[\"domain\"].append(domain)\n",
    "    domain_gene_table[\"gene_name\"].append(gene_name)\n",
    "    domain_gene_table[\"start\"].append(start)\n",
    "    domain_gene_table[\"end\"].append(end)\n",
    "    domain_gene_table[\"strand\"].append(strand)\n",
    "domain_gene_table = pd.DataFrame.from_dict(domain_gene_table)\n",
    "print(len(domain_gene_table))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['domain', 'gene_name', 'start', 'end', 'strand'], dtype='object')"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "domain_gene_table.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>domain</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Methyltransf_PK</td>\n",
       "      <td>CAH0700664.1</td>\n",
       "      <td>8507499</td>\n",
       "      <td>8508286</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Acetyltransf_1</td>\n",
       "      <td>CAH0686170.1</td>\n",
       "      <td>5403953</td>\n",
       "      <td>5405486</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Helicase_C</td>\n",
       "      <td>CAH0700515.1</td>\n",
       "      <td>5911986</td>\n",
       "      <td>5915269</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ARID</td>\n",
       "      <td>CAH0690184.1</td>\n",
       "      <td>8470044</td>\n",
       "      <td>8487486</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IKI3</td>\n",
       "      <td>CAH0687806.1</td>\n",
       "      <td>8046337</td>\n",
       "      <td>8054429</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>zf-H2C2_2</td>\n",
       "      <td>CAH0701113.1</td>\n",
       "      <td>4892528</td>\n",
       "      <td>4904033</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>SPRY</td>\n",
       "      <td>CAH0697809.1</td>\n",
       "      <td>6045211</td>\n",
       "      <td>6197982</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>WAC_Acf1_DNA_bd</td>\n",
       "      <td>CAH0690388.1</td>\n",
       "      <td>13982763</td>\n",
       "      <td>14004287</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>Glyco_transf_41</td>\n",
       "      <td>CAH0683039.1</td>\n",
       "      <td>4973789</td>\n",
       "      <td>4986696</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>Mst1_SARAH</td>\n",
       "      <td>CAH0694890.1</td>\n",
       "      <td>1133679</td>\n",
       "      <td>1154420</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>466 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              domain     gene_name     start       end strand\n",
       "0    Methyltransf_PK  CAH0700664.1   8507499   8508286      +\n",
       "1     Acetyltransf_1  CAH0686170.1   5403953   5405486      +\n",
       "2         Helicase_C  CAH0700515.1   5911986   5915269      +\n",
       "3               ARID  CAH0690184.1   8470044   8487486      +\n",
       "4               IKI3  CAH0687806.1   8046337   8054429      +\n",
       "..               ...           ...       ...       ...    ...\n",
       "461        zf-H2C2_2  CAH0701113.1   4892528   4904033      -\n",
       "462             SPRY  CAH0697809.1   6045211   6197982      +\n",
       "463  WAC_Acf1_DNA_bd  CAH0690388.1  13982763  14004287      -\n",
       "464  Glyco_transf_41  CAH0683039.1   4973789   4986696      -\n",
       "465       Mst1_SARAH  CAH0694890.1   1133679   1154420      -\n",
       "\n",
       "[466 rows x 5 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "domain_gene_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "domain_gene_table.to_csv(\"domain_gene_table.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "yx9C2WJSdk5N"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
